rust/lance/src/datafusion/logical_plan.rs

```diff
@@ -0,0 +1,144 @@
+        self
+    }
+
+    fn schema(&self) -> Arc<ArrowSchema> {
+        Arc::new(self.schema().into())
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+
+    fn get_table_definition(&self) -> Option<&str> {
+        None
+    }
+
+    fn get_logical_plan(&self) -> Option<&LogicalPlan> {
+        None
+    }
+
+    async fn scan(
+        &self,
+        _: &SessionState,
+        projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+        limit: Option<usize>,
+    ) -> DatafusionResult<Arc<dyn ExecutionPlan>> {
+        let scanner = self.scan();
+        let filter_expr = conjunction(filters.iter().cloned());
+        // NOTE: we only support node that have one partition. So any nodes that
+        // produce multiple need to be repartitioned to 1.
+        let mut filter_expr = if let Some(filter) = filter_expr {
+            let planner = Planner::new(Arc::new(scanner.dataset.schema().into()));
+            Some(planner.create_physical_expr(&filter)?)
+        } else {
+            None
+        };
+
+        // Stage 1: source
+        let mut plan: Arc<dyn ExecutionPlan> = if scanner.nearest.is_some() {
+            if scanner.prefilter {
+                let prefilter = filter_expr;
+                filter_expr = None;
+                scanner.knn(prefilter).await?
+            } else {
+                scanner.knn(None).await?
+            }
+        } else if let Some(expr) = filter_expr.as_ref() {
+            let columns_in_filter = column_names_in_expr(expr.as_ref());
+            let filter_schema = Arc::new(scanner.dataset.schema().project(&columns_in_filter)?);
+            scanner.scan(true, false, filter_schema)
+        } else {
+            // Scan without filter or limits
+            scanner.scan(
+                scanner.with_row_id,
+                false,
+                scanner.projections.clone().into(),
+            )
+        };
+
+        // Stage 2: filter
+        if let Some(predicates) = filter_expr.as_ref() {
+            let mut columns_in_filter = column_names_in_expr(predicates.as_ref());
+            // If we are going to sort then grab the ordering column at the same time we grab
+            // the columns we need for filtering
+            if let Some(ordering) = &scanner.ordering {
+                for column in ordering {
+                    if !columns_in_filter.contains(&column.column_name) {
+                        columns_in_filter.push(column.column_name.clone())
+                    }
+                }
+            }
+            let filter_schema = Arc::new(scanner.dataset.schema().project(&columns_in_filter)?);
+            let remaining_schema = filter_schema.exclude(plan.schema().as_ref())?;
+            if !remaining_schema.fields.is_empty() {
+                // Not all columns for filter are ready, so we need to take them first
+                plan = scanner.take(plan, &remaining_schema, scanner.batch_readahead)?;
+            }
```

Comment:
None of this is relevant. If we have a filter, then we already pulled all the columns in the original scan. And we aren't doing any sorting in this node, so we can ignore that part too.
```suggestion
```